meson.build

```diff
@@ -0,0 +1,362 @@
+# Copyright 2022 Collabora Ltd.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+project(
+  'Flatpak',
+  'c',
+  version : '1.13.2',
+  default_options: [
+    'warning_level=2',
+  ],
+  meson_version : '>=0.49.0',
+)
+
+# TODO: This is redundant with project->version
+flatpak_major_version = 1
+flatpak_minor_version = 13
+flatpak_micro_version = 2
+flatpak_extra_version = ''
+
+flatpak_version = '@0@.@1@.@2@@3@'.format(
+  flatpak_major_version,
+  flatpak_minor_version,
+  flatpak_micro_version,
+  flatpak_extra_version,
+)
+flatpak_interface_age = 0
+flatpak_binary_age = (
+  10000 * flatpak_major_version
+  + 100 * flatpak_minor_version
+  + flatpak_micro_version
+)
+
+required_glib = '2.44'
+required_bwrap = '0.5.0'
+required_dbus_proxy = '0.1.0'
+required_libostree = '2020.8'
+
+gnome = import('gnome')
+i18n = import('i18n')
+pkgconfig = import('pkgconfig')
+
+if meson.version().version_compare('>=0.56')
+  project_build_root = meson.project_build_root()
+  project_source_root = meson.project_source_root()
+else
+  project_build_root = meson.current_build_dir()
+  project_source_root = meson.current_source_dir()
+endif
+
+if meson.version().version_compare('>=0.58')
+  global_source_root = meson.global_source_root()
+else
+  global_source_root = meson.source_root()
+endif
+
+if meson.version().version_compare('>=0.53')
+  fs = import('fs')
+endif
+
+cc = meson.get_compiler('c')
+add_project_arguments('-include', 'config.h', language : 'c')
+common_include_directories = include_directories(
+  '.',
+  'common',
+)
+
+# Keep this in sync with ostree, except remove -Wall (part of Meson
+# warning_level 2) and -Werror=declaration-after-statement
+add_project_arguments(
+  cc.get_supported_arguments([
+    '-Werror=shadow',
+    '-Werror=empty-body',
+    '-Werror=strict-prototypes',
+    '-Werror=missing-prototypes',
+    '-Werror=implicit-function-declaration',
+    '-Werror=pointer-arith',
+    '-Werror=init-self',
+    '-Werror=missing-declarations',
+    '-Werror=return-type',
+    '-Werror=overflow',
+    '-Werror=int-conversion',
+    '-Werror=parenthesis',
+    '-Werror=incompatible-pointer-types',
+    '-Werror=misleading-indentation',
+    '-Werror=missing-include-dirs',
+
+    # Meson warning_level=2 would do this, but we are not fully
+    # signedness-safe yet
+    '-Wno-sign-compare',
+    '-Wno-error=sign-compare',
+
+    # Deliberately not warning about these, ability to zero-initialize
+    # a struct is a feature
+    '-Wno-missing-field-initializers',
+    '-Wno-error=missing-field-initializers',
+
+    # Deliberately not warning about these
+    '-Wno-unused-parameter',
+    '-Wno-error=unused-parameter',
+  ]),
+  language : 'c',
+)
+# Flatpak is Linux-specific, so for now we assume that -fvisibility=hidden
+# is always supported
+add_project_arguments('-fvisibility=hidden', language : 'c')
```

Comment:
You already target Meson 0.49, which means you can rely on `gnu_symbol_visibility: 'hidden'` as a kwarg to all relevant build targets.

(Granted it's safe to assume that on Linux all compilers support this. Even so -- Meson safely handles this by ignoring the kwarg on compilers that don't support it.)

Reply:
Thanks, I hadn't spotted that `gnu_symbol_visibility` existed.

Reply:
It turns out we need this anyway, until https://gitlab.gnome.org/GNOME/libglnx/-/merge_requests/34 lands.

Reply:
It's not clear to me why that is the case? add_project_arguments() won't affect subprojects any more than gnu_symbol_visibility on a target will.

Reply:
You're right, we need to fix this in libglnx (or work around it with a bigger hammer).